src/inmanta/server/protocol.py

```diff
@@ -384,8 +384,9 @@ def expire(self, timeout: float) -> None:
             IOLoop.current().remove_timeout(self._callhandle)
         self._sessionstore.expire(self, timeout)
 
-    def seen(self) -> None:
+    def seen(self, endpoint_names: List[str]) -> None:
         self._seen = time.time()
+        self.endpoint_names = endpoint_names
```

Comment:
This kind of complicates the agentmanager code, in the sense that you can not assume the session is not changed while you are handling it. (This change is outside the agentmanager session lock) 

I don't immediately know what the best way is to resolve this.

Reply:
(This code is surprisingly subtle, in that the session handling in the  SessionManager has no locks, but just has no await in crucial positions, while AgentManager has locks and awaits)

(Perhaps the AgentManager has to be split into multiple parts, as it is growing very big?)

Reply:
@wouterdb Currently we have two datastructures which maintain session state. One in the SessionManager and one in the AgentManager. Both have locking in place to make sure that their own datastructures stay consistent, but there is not locking across these two classes. So when a expire_session call and an create_session call happen at the same time in the SessionManager, they might be executed in reverse order on the AgentManager. This way both datastructures get out of sync. In my opinion the best way to solve this is, by using a fine-grained lock (on session id) in the SessionManager and stretch it towards the session listeners (AgentManager). What is your opinion on this? BTW, this problem also exists in the stable release, but probably the change of it being triggered is rather small.

Reply:
Let them be out-of-sync? I added some comments to that effect.

Reply:
However, a shared lock seems to become more and more attractive.